payjoin/src/receive/mod.rs

```diff
@@ -345,26 +346,46 @@ impl UncheckedProposal {
     }
 
     /// The Sender's Original PSBT
-    pub fn extract_tx_to_schedule_broadcast(&self) -> bitcoin::Transaction {
-        self.psbt.clone().extract_tx()
+    pub fn psbt_transaction(&self) -> bitcoin::Transaction { self.psbt.clone().extract_tx() }
+
+    /// The Sender's Original PSBT fee rate.
+    pub fn psbt_fee_rate(&self) -> Result<FeeRate, Error> {
+        if let Ok(original_psbt_fee) = self.psbt.fee() {
+            Ok(original_psbt_fee / self.psbt_transaction().weight())
+        } else {
+            Err(Error::BadRequest(InternalRequestError::OriginalPsbtNotBroadcastable.into()))
+        }
     }
 
-    /// Call after checking that the Original PSBT can be broadcast.
+    /// Check that the Original PSBT can be broadcasted.
     ///
     /// Receiver MUST check that the Original PSBT from the sender
-    /// can be broadcast, i.e. `testmempoolaccept` bitcoind rpc returns { "allowed": true,.. }
-    /// for `extract_tx_to_sheculed_broadcast()` before calling this method.
+    /// can be broadcast, i.e. `testmempoolaccept` bitcoind rpc returns { "allowed": true,.. }.
+    ///
+    /// Receiver can optionaly set a minimum feerate that will be enforced on the Original PSBT.
+    /// This can be used to prevent probing attacks and make it easier to deal with
+    /// high feerate environments.
     ///
     /// Do this check if you generate bitcoin uri to receive Payjoin on sender request without manual human approval, like a payment processor.
     /// Such so called "non-interactive" receivers are otherwise vulnerable to probing attacks.
     /// If a sender can make requests at will, they can learn which bitcoin the receiver owns at no cost.
     /// Broadcasting the Original PSBT after some time in the failure case makes incurs sender cost and prevents probing.
     ///
     /// Call this after checking downstream.
-    pub fn check_can_broadcast(
+    pub fn check_broadcast_suitability(
         self,
+        min_fee_rate: Option<FeeRate>,
         can_broadcast: impl Fn(&bitcoin::Transaction) -> Result<bool, Error>,
     ) -> Result<MaybeInputsOwned, Error> {
+        let original_psbt_fee_rate = self.psbt_fee_rate()?;
+        if let Some(min_fee_rate) = min_fee_rate {
+            if original_psbt_fee_rate < min_fee_rate {
+                return Err(Error::BadRequest(
+                    InternalRequestError::PsbtBelowFeeRate(original_psbt_fee_rate, min_fee_rate)
```

Comment:
I suggest implicit conversion to Error by using `.into()` on its own. See `return Err(InternalRequestError::InvalidContentType(content_type.to_owned()).into());`